receiver/snmpreceiver/scraper.go

```diff
@@ -0,0 +1,559 @@
+	// matchup later with returned SNMP indexed data
+	indexedMetricNamesByOID := map[string]string{}
+	indexedMetricOIDs := []string{}
+	for name, metricCfg := range s.cfg.Metrics {
+		if len(metricCfg.ColumnOIDs) > 0 {
+			for i, oid := range metricCfg.ColumnOIDs {
+				// Data is returned by the client with '.' prefix on the OIDs.
+				// Making sure the prefix exists here in the configs so we can match it up with returned data later
+				if !strings.HasPrefix(oid.OID, ".") {
+					oid.OID = "." + oid.OID
+					s.cfg.Metrics[name].ColumnOIDs[i].OID = oid.OID
+				}
+				indexedMetricOIDs = append(indexedMetricOIDs, oid.OID)
+				indexedMetricNamesByOID[oid.OID] = name
+			}
+		}
+	}
+
+	// If no column metric configs, nothing else to do
+	if len(indexedMetricOIDs) == 0 {
+		return nil
+	}
+
+	// Get all column OID SNMP indexed data for metrics and turn it into metrics, attributes,
+	// and resource attributes (using the previously retrieved attribute and resource attribute data)
+	err = s.client.GetIndexedData(
+		indexedMetricOIDs,
+		indexedDataToMetric(
+			now, resourceMetricsMap, metricsByResourceMap, indexedMetricNamesByOID,
+			indexedAttributeMapByOID, indexedResourceAttributeMapByOID, s,
+		),
+	)
+
+	return err
+}
+
+// indexedDataToMetric provides a function which will convert one piece of column OID SNMP indexed data
+// and turn it into a metric with attributes based on the config and previously collected column OID
+// SNMP indexed attribute. A resource may also be created if none exists for the related previously
+// collected resource attribute data yet.
+func indexedDataToMetric(
+	now pcommon.Timestamp,
+	resourceMetricsMap map[string]*pmetric.ResourceMetrics,
+	metricsByResourceMap map[string]map[string]*pmetric.Metric,
+	indexedMetricNamesByOID map[string]string,
+	indexedAttributeMapByOID map[string]map[string]string,
+	indexedResourceAttributeMapByOID map[string]map[string]string,
+	snmpScraper *snmpScraper,
+) processFunc {
+	// This returns a processFunc because this is what the client's GetScalarData method requires
+	return func(data snmpData) error {
+		// Return an error if this SNMP scalar data is not of a useable type
+		switch data.valueType {
+		case NotSupported:
+			fallthrough
+		case String:
+			return fmt.Errorf(errMsgIndexedBadValueType, data.oid, data.parentOID)
+		}
+
+		// Retrieve the metric config for this SNMP data
+		cfg := snmpScraper.cfg
+		metricName := indexedMetricNamesByOID[data.parentOID]
+		metricCfg := cfg.Metrics[metricName]
+
+		// Get all related attribute and resource attribute info for this SNMP indexed data based on the metric config
+		var metricResourceAttributes []string
+		var metricAttributes []Attribute
+		for _, columnOID := range metricCfg.ColumnOIDs {
+			if columnOID.OID == data.parentOID {
+				metricAttributes = columnOID.Attributes
+				metricResourceAttributes = columnOID.ResourceAttributes
+			}
+		}
```

Comment:
Same here can we create a lookup on the config values ahead of time rather than iterating every time?

Reply:
see my comment on a similar question